src/orbits/kepler.jl

```diff
@@ -1,110 +1,230 @@
+        t₀,
+        incl,
+        Ω,
+        ω,
+    )
+end
+
+@kwmethod function KeplerianOrbit(;aRₛ, b, ecc, P, t₀)
+    Ω = π / 2.0
+    ω = 0.0
+    incl = get_incl(aRₛ, b, ecc, sincos(ω))
+
+    return KeplerianOrbit(
+        nothing,
+        aRₛ,
+        b,
+        ecc,
+        P,
+        nothing,
+        nothing,
+        2.0 * π / P,
+        t₀,
+        incl,
+        Ω,
+        ω,
+    )
+end
+
+#############
+# Orbit logic
+#############
+# Star density
+get_ρₛ(aRₛ, P) = (3.0 * π / (G_val * P^2.0)) * aRₛ^3.0
+get_ρₛ(aRₛ, P::T) where {T <: Unitful.Time} = (3.0 * π / (G * P^2.0)) * aRₛ^3.0
+get_ρₛ(a, P, Rₛ) = get_ρₛ(aRₛ(a, Rₛ), P)
+
+# Semi-major axis / star radius ratio
+@kwdispatch get_aRₛ()
+@kwmethod get_aRₛ(;ρₛ, P) = cbrt(G_val * P^2.0 * ρₛ / (3.0 * π))
+@kwmethod get_aRₛ(;ρₛ, P::T) where {T <: Unitful.Time} = cbrt(G * P^2.0 * ρₛ / (3.0 * π))
+@kwmethod get_aRₛ(;a, P, Rₛ) = aRₛ(get_ρₛ(a, P, Rₛ), P)
+@kwmethod get_aRₛ(;a, Rₛ) = a / Rₛ
+
+# Semi-major axis
+get_a(ρₛ, P, Rₛ) = get_a(get_aRₛ(ρₛ=ρₛ, P=P), Rₛ)
+get_a(aRₛ, Rₛ) = aRₛ * Rₛ
+
+# Impact parameter
+get_b(ρₛ, P, sincosi) = get_b(get_aRₛ(ρₛ=ρₛ, P=P), sincosi)
+get_b(aRₛ, sincosi) = aRₛ * sincosi[2]
+
+# Inclination
+function get_incl(aRₛ, b, ecc, sincosω)
+    return acos((b/aRₛ) * (1.0 + ecc*sincosω[1])/(1.0 - ecc^2))
 end
 
+# Finds the position `r` of the planet along its orbit after rotating
+# through the true anomaly `ν`, then transforms this from the
+# orbital plan to the equatorial plane
 function relative_position(orbit::KeplerianOrbit, t)
-    sinf, cosf = get_true_anomaly(orbit, t)
+    sinν, cosν = get_true_anomaly(orbit, t)
     if orbit.ecc === nothing
-        r = -orbit.a
+        r = orbit.aRₛ
     else
-        r = -orbit.a * (1 - orbit.ecc^2) / (1 + orbit.ecc * cosf)
+        r = orbit.aRₛ * (1 - orbit.ecc^2) / (1 + orbit.ecc * cosν)
     end
-    return rotate_vector(orbit, r * cosf, r * sinf)
+    return rotate_vector(orbit, r * cosν, r * sinν)
 end
 
-function get_true_anomaly(orbit::KeplerianOrbit, t)
-    M = orbit.n * ((t - orbit.t0) - orbit.tref)
-    return kepler(M, orbit.ecc)
+function get_M(orbit::KeplerianOrbit, t)
+    if orbit.ecc == 0.0
+        M = M₀ + orbit.n * (t - orbit.t₀) / 2.0
+    else
+        sinω, cosω = sincos(orbit.ω)
+        E₀ = 2.0 * atan(
+            sqrt(1.0 - orbit.ecc) * cosω,
+            sqrt(1.0 + orbit.ecc) * (1.0 + sinω),
+        )
+        M₀_ecc = E₀ - orbit.ecc * sin(E₀)
+        M = M₀_ecc + orbit.n * (t - orbit.t₀)
+    end
+    return M
 end
 
-function kepler(M, ecc)
-    E = kepler_solver(M, ecc)
-    return sincos(E)
+# Returns sin(ν), cos(ν)
+function get_true_anomaly(orbit::KeplerianOrbit, t)
+    #M = get_M(orbit, t)
+    M = M₀ + orbit.n * (t - orbit.t₀) / 2.0
+    E = kepler_solver(M, orbit.ecc)
+    return sincos(trueanom(E, orbit.ecc))
 end
-kepler(M, ::Nothing) = sincos(M)
 
+# Transform from orbital plane to equatorial plane
 function rotate_vector(orbit::KeplerianOrbit, x, y)
```

Comment:
If you want to play around with this, you could also see if you like [Rotations.jl](https://github.com/juliageometry/Rotations.jl). It may end up with similar logic, but it's your discretion.

Reply:
Oh neat